virtcontainers/clh.go

```diff
@@ -373,8 +375,29 @@ func (clh *cloudHypervisor) getThreadIDs() (vcpuThreadIDs, error) {
 }
 
 func (clh *cloudHypervisor) hotplugAddDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
-	clh.Logger().WithField("function", "hotplugAddDevice").Warn("hotplug add device not supported")
-	return nil, nil
+	span, _ := clh.trace("hotplugAddDevice")
+	defer span.Finish()
+
+	switch devType {
+	case vfioDev:
+		device := devInfo.(*config.VFIODev)
+
+		cl := clh.client()
+		ctx, cancel := context.WithTimeout(context.Background(), clhHotPlugAPITimeout*time.Second)
+		defer cancel()
+
+		_, _, err := cl.VmmPingGet(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = cl.VmAddDevicePut(ctx, chclient.VmAddDevice{Path: device.SysfsDev})
+
+		return nil, openAPIClientError(err)
```

Comment:
You may want to add more information to the `err` being returned if `openAPIClientError()` won't tell which API is failing. I found it is easier to trace the errors related to CLH API calls.

Reply:
Added more information :)